protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,713 @@
+    }
+
+    fn inject_expired_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_expired_listen_addr(id, addr);
+    }
+
+    fn inject_listener_error(&mut self, id: ListenerId, err: &(dyn std::error::Error + 'static)) {
+        self.inner.inject_listener_error(id, err)
+    }
+
+    fn inject_listener_closed(&mut self, id: ListenerId, reason: Result<(), &std::io::Error>) {
+        self.inner.inject_listener_closed(id, reason)
+    }
+
+    fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_new_external_addr(addr);
+        if self.reachability.is_public() || self.pending_probe.is_some() {
+            return;
+        }
+        if let Some((ref auto_retry, _)) = self.auto_retry {
+            let probe = self.new_probe(
+                auto_retry.required_success,
+                Vec::new(),
+                true,
+                auto_retry.extend_with_connected,
+                auto_retry.max_peers,
+                auto_retry.min_valid,
+            );
+            self.pending_probe.replace(probe);
+        }
+    }
+
+    fn inject_expired_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_expired_external_addr(addr);
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        if let Some((ref auto_retry, ref mut delay)) = self.auto_retry {
+            if delay.poll_unpin(cx).is_ready()
+                && self.ongoing_outbound.is_none()
+                && self.pending_probe.is_none()
+            {
+                let probe = self.new_probe(
+                    auto_retry.required_success,
+                    Vec::new(),
+                    true,
+                    auto_retry.extend_with_connected,
+                    auto_retry.max_peers,
+                    auto_retry.min_valid,
+                );
+                let records: Vec<_> = params.external_addresses().collect();
+                self.do_probe(probe, records);
+            }
+        };
+        if let Some(probe) = self.pending_probe.take() {
+            let records: Vec<_> = params.external_addresses().collect();
+            self.do_probe(probe, records);
+        }
+        loop {
+            if let Some(event) = self.pending_out_event.pop_front() {
+                return Poll::Ready(NetworkBehaviourAction::GenerateEvent(event));
+            }
+            match self.inner.poll(cx, params) {
+                Poll::Ready(NetworkBehaviourAction::GenerateEvent(event)) => match event {
+                    RequestResponseEvent::Message { peer, message } => match message {
+                        RequestResponseMessage::Request {
+                            request_id: _,
+                            request,
+                            channel,
+                        } => match self.handle_request(peer, request) {
```

Comment:
These _could_ be collapsed:

```suggestion
                Poll::Ready(NetworkBehaviourAction::GenerateEvent(RequestResponseEvent::Message { peer, message: RequestResponseMessage::Request {
                            request_id: _,
                            request,
                            channel,
                        } })) => match self.handle_request(peer, request) {
```

I don't know what rustfmt will make out of this but it should reduce the indentation level for all the following code a bit due to the removed nested `match`s.